src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateNow = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateNow, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateNow[key];
+        }
+        break;
+
+      case 'clear':
+        Object.keys(stateNow).forEach((key) => delete stateNow[key]);
+        break;
+
+      default:
+        break;
+    }
+
+    states.push({ ...stateNow });
```

Comment:
It's good that you're creating a new object copy before pushing it to the 'states' array to avoid mutating the original state. However, this should be done after the switch statement within the loop to ensure that the action is fully processed before creating the clone.